src/diffusers/models/autoencoders/autoencoder_kl_allegro.py

```diff
@@ -0,0 +1,1024 @@
+            hidden_states = self.temp_convs_down(hidden_states)
+
+        if self.downsamplers is not None:
+            hidden_states = hidden_states.permute(0, 2, 1, 3, 4).flatten(0, 1)
+            for downsampler in self.downsamplers:
+                hidden_states = downsampler(hidden_states)
+            hidden_states = hidden_states.unflatten(0, (batch_size, -1)).permute(0, 2, 1, 3, 4)
+
+        return hidden_states
+
+
+class AllegroUpBlock3D(nn.Module):
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        dropout: float = 0.0,
+        num_layers: int = 1,
+        resnet_eps: float = 1e-6,
+        resnet_time_scale_shift: str = "default",  # default, spatial
+        resnet_act_fn: str = "swish",
+        resnet_groups: int = 32,
+        resnet_pre_norm: bool = True,
+        output_scale_factor=1.0,
+        add_upsample=True,
+        add_temp_upsample=False,
+        temb_channels=None,
+    ):
+        super().__init__()
+        self.add_upsample = add_upsample
+
+        resnets = []
+        temp_convs = []
+
+        for i in range(num_layers):
+            input_channels = in_channels if i == 0 else out_channels
+
+            resnets.append(
+                ResnetBlock2D(
+                    in_channels=input_channels,
+                    out_channels=out_channels,
+                    temb_channels=temb_channels,
+                    eps=resnet_eps,
+                    groups=resnet_groups,
+                    dropout=dropout,
+                    time_embedding_norm=resnet_time_scale_shift,
+                    non_linearity=resnet_act_fn,
+                    output_scale_factor=output_scale_factor,
+                    pre_norm=resnet_pre_norm,
+                )
+            )
+            temp_convs.append(
+                AllegroTemporalConvBlock(
+                    out_channels,
+                    out_channels,
+                    dropout=0.1,
+                )
+            )
+
+        self.resnets = nn.ModuleList(resnets)
+        self.temp_convs = nn.ModuleList(temp_convs)
+
+        self.add_temp_upsample = add_temp_upsample
+        if add_temp_upsample:
+            self.temp_conv_up = AllegroTemporalConvBlock(
+                out_channels, out_channels, dropout=0.1, up_sample=True, stride=3
+            )
+
+        if self.add_upsample:
+            self.upsamplers = nn.ModuleList([Upsample2D(out_channels, use_conv=True, out_channels=out_channels)])
+        else:
+            self.upsamplers = None
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        batch_size = hidden_states.shape[0]
+
+        for resnet, temp_conv in zip(self.resnets, self.temp_convs):
+            hidden_states = hidden_states.permute(0, 2, 1, 3, 4).flatten(0, 1)
+            hidden_states = resnet(hidden_states, temb=None)
+            hidden_states = hidden_states.unflatten(0, (batch_size, -1)).permute(0, 2, 1, 3, 4)
```

Comment:
I think the current implementation is much cleaner, no? Maybe we could do a refactor of the existing `TemporalConvLayer` in a future PR? Either okay with me, so just need your confirmation before committing to this (as there are many instances of this and using 3 lines instead of 1 will look the code look a bit more bloatified)

Reply:
Or did you mean to make this change just in the temporal layer, while leaving the other instances of unflatten/permute untouched?

```python
        hidden_states = (
            hidden_states[None, :].reshape((-1, num_frames) + hidden_states.shape[1:]).permute(0, 2, 1, 3, 4)
        )
```

Reply:
yeah I just glance through the code, I think this line  `hidden_states.unflatten(0, (batch_size, -1)).permute(0, 2, 1, 3, 4)` is everywhere, and basically everytime you call the temp conv layer, you need to do that - so it is easier to just process inside the temp conv I think  https://github.com/huggingface/diffusers/pull/9736/files#r1809793907

Reply:
will do next, thanks for the review! this was just a 1:1 conversion of the original code using einops so i didn't try to get clever and break anything